runtime/vm/LayoutFFITypeHelpers.hpp

```diff
@@ -0,0 +1,270 @@
+		case 'V': /* VOID */
+			typeFFI = &ffi_type_void;
+			break;
+		case 'B': /* C_BOOL */
+			typeFFI = &ffi_type_uint8;
+		case 'C': /* C_CHAR */
+			typeFFI = &ffi_type_sint8;
+			break;
+		case 'S': /* C_SHORT */
+			typeFFI = &ffi_type_sint16;
+			break;
+		case 'I': /* C_INT */
+			typeFFI = &ffi_type_sint32;
+			break;
+		case 'J': /* 8 bytes for either C_LONG (Linux) or C_LONG_LONG(maps to long long specific to Windows/AIX 64-bit) */
+			typeFFI = &ffi_type_sint64;
+			break;
+		case 'F': /* C_FLOAT */
+			typeFFI = &ffi_type_float;
+			break;
+		case 'D': /* C_DOUBLE */
+			typeFFI = &ffi_type_double;
+			break;
+		case 'P': /* C_POINTER */
+			typeFFI = &ffi_type_pointer;
+			break;
+		default:
+			Assert_VM_unreachable();
+		}
+
+		return typeFFI;
+	}
+
+	/**
+	 * @brief Create a FFI type from a layout string object for both primitive and struct
+	 *
+	 * @param typeFFI[in] The custom FFI type to be created
+	 * @param layoutStringObject[in] An object containing the layout string describing the field types.
+	 * @return The size of the layout, extracted from the layout string, UDATA_MAX if a failure occurs
+	 */
+	VMINLINE UDATA
+	getLayoutFFIType(ffi_type** typeFFI, j9object_t layoutStringObject)
+	{
+		PORT_ACCESS_FROM_JAVAVM(_vm);
+		*typeFFI = NULL;
+		UDATA layoutSize = 0;
+		char layoutBuffer[J9VM_LAYOUT_STRING_ON_STACK_LIMIT];
+		char *layout = copyStringToUTF8WithMemAlloc(_currentThread, layoutStringObject,
+				J9_STR_NULL_TERMINATE_RESULT, "", 0, layoutBuffer, sizeof(layoutBuffer), NULL);
+		/* Preserve the original pointer for j9mem_free_memory() as
+		 * subsequent calls will modify the contents of this pointer.
+		 */
+		char *layoutTemp = layout;
+		if (NULL == layout) {
+			goto done;
+		}
+
+		/* Check the byte size of the layout's size which is prefixed to the layout string */
+		layoutSize = getIntFromLayout(&layoutTemp);
+		if (layoutSize >= UDATA_MAX) {
+			return layoutSize;
+		}
+
+		/* There are 2 kinds of layout: ValueLayout for primitive and GroupLayout for struct.
+		 * e.g.
+		 * 1) int: 4I  (4 represents the int layout's size in bytes)
+		 * 2) struct:
+		 *    struct struct_III{
+		 *                       int elem1;
+		 *                       int elem2;
+		 *                       int elem3;
+		 *                     };
+		 *    with the layout (12#3[III]) as follows:
+		 *    12#3[   (12 represents the struct layout's size in bytes while 3 represents the counts of elements)
+		 *         I  (C_INT)
+		 *         I  (C_INT)
+		 *         I  (C_INT)
+		 *        ]
+		 *    where the header of the struct layout is "12#3".
+		 * Thus, a struct layout string starts with the layout's size, a '#' as separator
+		 * plus the count of struct elements as the header.
+		 *
+		 * Note: the layout string is preprocessed and prefixed with the header in
+		 * ProgrammableInvoker.preprocessLayoutString() for easier handling in native.
+		 */
+		if ('#' == *layoutTemp) {
+			*typeFFI = getStructFFIType(&layoutTemp);
```

Comment:
we should make all struct types singletons, similar to who the primitives are singletons.

Reply:
I don't quite follow this as struct types are quite different from primitives (will be shared once created) given each struct type might be mixed with different layouts which can't be shared unless they are totally identical. Actually we already cache these types in java if equal, in which case there is no need to create a new struct type if they are equal.

Reply:
what I mean is if we have two structs
```
struct A {
    int a1;
    int a2;
}

struct Point {
    int x;
    int y;
}
```

We should return the same ffitype for both rather than making two new ones.

Reply:
For the same method signature, we already have `cachedArgTypes` in Java to avoid creating the same ffitypes. But if it means to check each argument or even each basic struct type nested in a struct, things become totally different.

The question for this is what depth we really pursue in terms of ffitype, which could be more challenging when it comes to a nested struct type because the parsing of struct type in native is expend recursively in a liner way in which case there is no easy way to determine whether they are identical in advance unless we do that to in Java by caching these created ffitype. But that means we have to cache the ffitype in native and the key is how to determine the same ffitype is cached. 
e.g.
```
struct A {
   struct Point x,
   struct Point  y,
}
```
How to determine whether to return the same ffitype for `y` in native after creating a ffitype for `x` ? (the actual situations could be more complex than this case.

Reply:
The only way I can image is to cache each created stuct ffitype in native (rather than in java) when creating a new one, and compare it with the existing one in the cache (probably by checking the hash of the ffitype). But the question still remains: we might need to know exactly each element of this stuct type in native so as to compute the hash for that.

Reply:
Considering the complexity for the proposed solution, we might need to create a separate issue if we really do optimization and get back to figure how to get this issue addressed after implementing everything (downcall & upcall) in JEP389/419.

Reply:
We can pass on this for now